akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/BehaviorSetup.scala

```diff
@@ -102,8 +103,16 @@ private[akka] final class BehaviorSetup[C, E, S](
     recoveryTimer = OptionVal.None
   }
 
-  def onSignal(signal: Signal): Unit = {
-    signalHandler.applyOrElse(signal, ConstantFun.scalaAnyToUnit)
+  def onSignal(state: S, signal: Signal): Unit = {
+    try {
+      signalHandler.applyOrElse((state, signal), ConstantFun.scalaAnyToUnit)
+    } catch {
+      case NonFatal(ex) =>
+        if (signal == akka.persistence.typed.RecoveryCompleted)
+          throw ex // fail the recovery
+        else
+          log.error(ex, s"Error while processing signal [{}]", signal)
```

Comment:
I know this was existing behavior but: shouldn't it always fail if a signal handler throws?

Reply:
Good point, I don't see a reason to catch exceptions here. A signal handler should work in the same way as a command handler (or side effect callback). Exception should be handled by supervision when it has reached the Running phase.

Before Running, exception from RecoveryCompleted should fail the recovery (stopping the actor unconditionally). Exceptions from other signals before Running should probably be caught and logged like this to avoid restart loops.

Reply:
@johanandren I have improved this, take a look at d011f0a

Reply:
Good!